optuna/samplers/_ga/_base.py

```diff
@@ -0,0 +1,180 @@
+    :meth:`~BaseGASampler.get_generation` and :meth:`~BaseGASampler.get_population`, ensuring
+    consistent tracking and selection.
+
+    Note:
+        This class should be extended by subclasses that define specific GA sampling strategies,
+        including parent selection and crossover operations.
+    """
+
+    _GENERATION_KEY = "BaseGASampler:generation"
+    _PARENT_CACHE_KEY_PREFIX = "BaseGASampler:parent:"
+
+    def __init_subclass__(cls, **kwargs: Any):
+        super().__init_subclass__(**kwargs)
+        cls._GENERATION_KEY = f"{cls.__name__}:generation"
+        cls._PARENT_CACHE_KEY_PREFIX = f"{cls.__name__}:parent:"
+
+    @classmethod
+    def _get_generation_key(cls) -> str:
+        return cls._GENERATION_KEY
+
+    @classmethod
+    def _get_parent_cache_key_prefix(cls) -> str:
+        return cls._PARENT_CACHE_KEY_PREFIX
+
+    def __init__(self, population_size: int):
+        self._population_size = population_size
+
+    @property
+    def population_size(self) -> int:
+        return self._population_size
+
+    @population_size.setter
+    def population_size(self, value: int) -> None:
+        self._population_size = value
+
+    @abc.abstractmethod
+    def select_parent(self, study: optuna.Study, generation: int) -> list[FrozenTrial]:
+        """Select parent trials from the population for the given generation.
+
+        This method is called once per generation to select parents from
+        the population of the current generation.
+
+        Output of this function is cached in the study system attributes.
+
+        This method must be implemented in a subclass to define the specific selection strategy.
+
+        Args:
+            study:
+                Target study object.
+            generation:
+                Target generation number.
+
+        Returns:
+            List of parent frozen trials.
+        """
+        raise NotImplementedError
+
+    def get_generation(self, study: optuna.Study, trial: FrozenTrial) -> int:
+        """Get the current generation number of the given trial.
+
+        This method returns the generation number of the specified trial. If the generation number
+        is not set in the trial's system attributes, it will calculate and set the generation
+        number.
+
+        The current generation number is defined as the maximum generation number among all
+        completed trials.
+
+        Args:
+            study:
+                Study object which trial belongs to.
+            trial:
+                Trial object to get the generation number.
+
+        Returns:
+            Generation number of the given trial.
+        """
+        generation = trial.system_attrs.get(self._get_generation_key(), None)
+        if generation is not None:
+            return generation
+
+        trials = study._get_trials(deepcopy=False, states=[TrialState.COMPLETE], use_cache=True)
+
+        max_generation, max_generation_number = 0, 0
```

Comment:
I prefer `max_generation_individuals_count` or something easy to understand to `max_generation_number` based on my understanding that this variable indicates the number of individuals belonging to the maximum generation.